proxy/Dockerfile

```diff
@@ -1,24 +1,106 @@
 # Proxy build and runtime
 #
-# Builds a slim runtime image with the conduit-proxy binary.
+# When BUILD_UNOPTIMIZED is set and not empty, unoptimized rust artifacts are produced.
+# This reduces build time and produces binaries with debug symbols, at the expense of
+# runtime perforamnce.
+#
+# When SKIP_TESTS is set and not empty, tests are not run. Otherwise, tests are run
+# against either unoptimized or optimized proxy code, according to BUILD_UNOPTIMIZED.
+
+ARG RUST_IMAGE=rust:1.23.0
+
+ARG RUNTIME_IMAGE=gcr.io/runconduit/base:2017-10-30.01
+
+## Fetch external dependencies.
+#
+# Mocks out all local code and fetch external dependencies to ensure that external sources
+# are cached.
+FROM $RUST_IMAGE as fetch
```

Comment:
not sure how actionable this is, or how much it will affect build performance, but these multistage builds come at some cost to disk space:
```bash
$ docker images
REPOSITORY                                             TAG                  IMAGE ID            CREATED             SIZE
gcr.io/runconduit/proxy                                git-03261aaf         c0a0a95e20f9        6 minutes ago       109 MB
<none>                                                 <none>               aa25ba002caa        6 minutes ago       2.02 GB
<none>                                                 <none>               a59dda926066        44 minutes ago      1.93 GB
<none>                                                 <none>               064aba0af483        46 minutes ago      1.88 GB
<none>                                                 <none>               ee978ff534a7        50 minutes ago      1.78 GB
```

Reply:
At least we're no longer trying to ship these intermediates around the network ;)

Reply:
It does seem unfortunate that there is so much disk space being used but I think we can improve that later.